File: typescript/src/linked-list/list.ts
export type List = ListNode | null;

export class ListNode {
  value: number;
  next: List;

  constructor(value = 0, next: List = null) {
    this.value = value;
    this.next = next;
  }

  static from_values(values: Iterable<number>): List {
    let head: List = null;
    let cursor: List = null;
    for (const val of values) {
      const node = new ListNode(val);
      if (!cursor) {
        head = node;
      } else {
        cursor.next = node;
      }
      cursor = node;
    }
    return head;
  }

  *nodes(): Generator<ListNode> {
    let cursor: List = this;
    while (cursor) {
      yield cursor;
      cursor = cursor.next;
    }
  }

  *values(): Generator<number> {
    for (const node of this.nodes()) {
      yield node.value;
    }
  }
}

File: python/src/linked_list/list.py
from __future__ import annotations
from typing import Iterable, Iterator, Optional


class ListNode:
    def __init__(self, value: int = 0, next: List = None):
        self.value = value
        self.next = next

    @classmethod
    def from_values(cls, values: Iterable[int]) -> List:
        head = None
        cursor = None
        for value in values:
            node = cls(value)
            if cursor is None:
                head = node
            else:
                cursor.next = node
            cursor = node
        return head

    def nodes(self) -> Iterator[ListNode]:
        cursor: List = self
        while cursor:
            yield cursor
            cursor = cursor.next

    def values(self) -> Iterator[int]:
        return map(lambda node: node.value, self.nodes())


List = Optional[ListNode]
